File: selenium-project/src/main/java/com/master/seleniumproject/pages/books/BookForm.java
package com.master.seleniumproject.pages.books;

import com.master.seleniumproject.pages.AbstractPage;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;

public class BookForm extends AbstractPage  {

    @FindBy(css = "input[id='title']")
    private WebElement titleInput;

    @FindBy(css = "input[id='author']")
    private WebElement authorInput;

    @FindBy(css = "input[id='genre']")
    private WebElement genreInput;

    @FindBy(css = "input[id='year']")
    private WebElement yearInput;

    @FindBy(css = "button[id='add-update-form-button']")
    private WebElement addUpdateButton;

    @FindBy(css = "button[id='cancel-button']")
    private WebElement cancelButton;

    public BookForm(WebDriver driver) {
        super(driver);
        PageFactory.initElements(driver, this);
    }

    public void fillTtitleInput(String ttitle) {
        this.fillInput(this.titleInput, ttitle);
    }

    public void fillAuthorInput(String author) {
        this.fillInput(this.authorInput, author);
    }

    public void fillGenreInput(String genre) {
        this.fillInput(this.genreInput, genre);
    }

    public void fillYearInput(String year) {
        this.fillInput(this.yearInput, year);
    }

    public void clickAddUpdateButton() {
        this.addUpdateButton.click();
    }

    public void clickCancelButton() {
        this.cancelButton.click();
    }
}

File: playwright-project/src/pages/books/bookForm.ts
import {Locator, Page} from '@playwright/test';
import {BasePage} from '@pages/basePage';

export class BookForm extends BasePage {
    constructor(
        page: Page,
        private titleInput: Locator = page.locator(`input[id='title']`),
        private authorInput: Locator = page.locator(`input[id='author']`),
        private genreInput: Locator = page.locator(`input[id='genre']`),
        private yearInput: Locator = page.locator(`input[id='year']`),
        private addUpdateButton: Locator = page.locator(`button[id='add-update-form-button']`),
        private cancelButton: Locator = page.locator(`button[id='cancel-button']`),
    ) {
        super(page);
    }

    async openLoginPage(loginPageUrl: string): Promise<void> {
        await this.openURL(loginPageUrl);
    }

    async fillTitle(title: string): Promise<void> {
        await this.titleInput.fill(title);
    }

    async fillAuthor(author: string): Promise<void> {
        await this.authorInput.fill(author);
    }

    async fillGenre(genre: string): Promise<void> {
        await this.genreInput.fill(genre);
    }

    async fillYear(year: string): Promise<void> {
        await this.yearInput.fill(year);
    }

    async clickAddUpdateButton(): Promise<void> {
        await this.addUpdateButton.click();
    }

    async clickCancelButton(): Promise<void> {
        await this.cancelButton.click();
    }
}
